sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBCheckpointFailureInjectionSuite.scala

```diff
@@ -0,0 +1,560 @@
+            checkpointId2 = commitAndGetCheckpointId(db)
+
+            assert(FailureInjectionFileSystem.delayedStreams.nonEmpty)
+            FailureInjectionFileSystem.delayedStreams.foreach(_.close())
+
+            db.load(1, checkpointId1)
+            db.load(2, checkpointId2)
+          }
+
+          withDB(
+            remoteDir.getAbsolutePath,
+            version = 2,
+            conf = conf,
+            hadoopConf = hadoopConf,
+            enableStateStoreCheckpointIds = ifEnableStateStoreCheckpointIds,
+            checkpointId = checkpointId2) { db =>
+            if (ifEnableStateStoreCheckpointIds) {
+              assert(new String(db.get("version"), "UTF-8") == "2.2")
+            } else {
+              // This check is not necessary. But we would like to validate the behavior of
+              // FailureInjectionFileSystem to ensure the ifEnableStateStoreCheckpointIds = true
+              // case is valid.
+              assert(new String(db.get("version"), "UTF-8") == "2.1")
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * This test is to simulate the case where
+   * 1. There is a snapshot checkpoint scheduled
+   * 2. The batch eventually failed
+   * 3. Query is retried and moved forward
+   * 4. The snapshot checkpoint succeeded
+   * In checkpoint V2, this snapshot shouldn't take effect. Otherwise, it will break the strong
+   * consistency guaranteed by V2.
+   */
+  test("Delay Snapshot V2") {
+    val fmClass = "org.apache.spark.sql.execution.streaming.state." +
+      "FailureInjectionCheckpointFileManager"
+    val hadoopConf = new Configuration()
+    hadoopConf.set(STREAMING_CHECKPOINT_FILE_MANAGER_CLASS.parent.key, fmClass)
+    withTempDir { remoteDir =>
+      withSQLConf(
+        RocksDBConf.ROCKSDB_SQL_CONF_NAME_PREFIX + ".changelogCheckpointing.enabled" -> "true",
+        SQLConf.STATE_STORE_MIN_DELTAS_FOR_SNAPSHOT.key -> "2") {
+        val conf = RocksDBConf(StateStoreConf(SQLConf.get))
+        var checkpointId3: Option[String] = None
+        withDB(
+          remoteDir.getAbsolutePath,
+          version = 0,
+          conf = conf,
+          hadoopConf = hadoopConf,
+          enableStateStoreCheckpointIds = true) { db =>
+          db.put("version", "1.1")
+          val checkpointId1 = commitAndGetCheckpointId(db)
+
+          withDB(
+            remoteDir.getAbsolutePath,
+            version = 1,
+            conf = conf,
+            hadoopConf = hadoopConf,
+            enableStateStoreCheckpointIds = true,
+            checkpointId = checkpointId1) { db2 =>
+            db2.put("version", "2.1")
+            db2.commit()
+
+            db.load(1, checkpointId1)
+            db.put("version", "2.2")
+            val checkpointId2 = commitAndGetCheckpointId(db)
+
+            db.load(2, checkpointId2)
+            db.put("foo", "bar")
+            checkpointId3 = commitAndGetCheckpointId(db)
+
+            db2.doMaintenance()
```

Comment:
Do we have a test with varying maintenance intervals ? i.e. where maintenance is configured to run very often vs very rarely ? maybe not for each case - but for some of the most common test cases here ?

Reply:
We don't have it. Maybe do it as a follow up when we have time later.